velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -1042,6 +1044,123 @@ TpchPlan TpchQueryBuilder::getQ19Plan() const {
+  const auto& partsuppFileColumns = getFileColumnNames(kPartsupp);
+  auto nationSelectedRowType = getRowType(kNation, nationColumns);
+  const auto& nationFileColumns = getFileColumnNames(kNation);
+
+  const std::string shipDateFilter = formatDateFilter(
+      "l_shipdate", lineitemSelectedRowType, "'1994-01-01'", "'1994-12-31'");
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemScanNodeId;
+  core::PlanNodeId partScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId partsuppScanNodeId;
+  core::PlanNodeId nationScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator)
+                  .tableScan(
+                      kPart,
+                      partSelectedRowType,
+                      partFileColumns,
+                      {},
+                      "p_name like 'forest%'")
+                  .capturePlanNodeId(partScanNodeId)
+                  .planNode();
+
+  auto nation = PlanBuilder(planNodeIdGenerator)
+                    .tableScan(
+                        kNation,
+                        nationSelectedRowType,
+                        nationFileColumns,
+                        {"n_name = 'CANADA'"})
+                    .capturePlanNodeId(nationScanNodeId)
+                    .planNode();
+
+  auto partsuppJoinPart =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kPartsupp, partsuppSelectedRowType, partsuppFileColumns)
+          .capturePlanNodeId(partsuppScanNodeId)
+          .hashJoin(
+              {"ps_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"ps_partkey", "ps_suppkey", "ps_availqty"},
+              core::JoinType::kLeftSemi)
+          .planNode();
+
+  auto lineitemJoinPartsupp =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kLineitem,
+              lineitemSelectedRowType,
+              lineitemFileColumns,
+              {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .partialAggregation(
+              {"l_partkey", "l_suppkey"}, {"sum(l_quantity) AS sum_qty"})
+          .localPartition({})
+          .finalAggregation()
+          .project({"l_partkey", "l_suppkey", "0.5 * sum_qty AS filter_qty"})
+          .hashJoin(
+              {"l_partkey", "l_suppkey"},
+              {"ps_partkey", "ps_suppkey"},
+              partsuppJoinPart,
+              "ps_availqty > filter_qty",
+              {"ps_suppkey"})
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kSupplier, supplierSelectedRowType, supplierFileColumns)
+          .capturePlanNodeId(supplierScanNodeId)
+          .hashJoin(
+              {"s_nationkey"},
+              {"n_nationkey"},
+              nation,
+              "",
+              {"s_name", "s_address", "s_suppkey"})
+          // A left semi-join is used here to keep the query plan consistent
```

Comment:
Why do we want to keep the plan consistent with DuckDB? Is this the best plan? If not, let's use the best plan.

Reply:
Since the comparison is with DuckDb, we are trying to adhere to the DuckDb plan. However, if an ideal optimizer is expected to generate a better plan, we follow the optimal plan in that case. Here, DuckDb query plan is optimal so it is followed. Once right semi-join support is added to Velox, the plan can be modified.

Reply:
> if an ideal optimizer is expected to generate a better plan, we follow the optimal plan in that case

Let's then remove the comment as it suggests that we prefer DuckDB plan over optimal plan. Let's also add a comment to explain that it would be better to use a right semi-join, but it is not supported yet.

Reply:
Sure, thanks

Reply:
The anomaly in time taken between 8 and 16 threads can be attributed to the following stages:
```
| Step               | Time taken: 8 threads | Time taken: 16 threads |  
|--------------------|-----------------------|------------------------|
| Lineitem tablescan |                  8.32 |                   9.92 |
| PartialAgg sum_qty |                  8.19 |                    9.8 | 
| FinalAgg sum_qty   |                  3.43 |                   3.95 |
| Total time         |                  2.82 |                   3.02 |
```

Reply:
@pramodsatya Would you take a look into what's causing the slowness?

Reply:
Sure, thanks

Reply:
Hi @mbasmanova, apologies for the delay in looking into this. From the blocked times for tablescans, joins, and aggregations in the query, the join between lineitem and partsupp on ```suppkey, partkey``` accounts for the slowness seen with 16 threads. The table below represents the blocked time at the build side of this hash join. 
```
| Num threads | Time (in ms) |
|:-----------:|:------------:|
|      1      |       0      |
|      4      |     69.86    |
|      8      |    200.43    |
|      16     |    861.53    |
```
The hash table recombination on the build side following the partitioning of the data on partkey, suppkey (for the ```filter_qty``` aggregation) blocks the hash join. This blocked time increases with the number of threads as the partitioned data at each thread needs to be merged to a single hash table at the build side before the join can proceed. Since the benchmarks are run on a 6 core i7 machine, the difference between 8 and 16 threads is significant (~0.7s) compared to that between 4 and 8 threads. This accounts for the slowness seen with 16 threads compared to 8 threads. The tablescan times reported earlier were CPU times and do not directly account for the wallclock times, apologies for the confusion.